pgbot/commands/utils/browse.py

```diff
@@ -0,0 +1,120 @@
+import discord
+
+from pgbot import common
+from pgbot.commands.base import BotException
+from pgbot.utils import embed_utils
+
+
+class MessageBrowser(embed_utils.PagedEmbed):
+    def __init__(self, message, caller=None, command=None, start_page=0):
+        super().__init__(
+            message,
+            [discord.embeds.EmptyEmbed] * 3,
+            caller=caller,
+            command=command,
+            start_page=start_page,
+        )
+
+    async def setup(self, channel, quantity, before=None, after=None, around=None):
+        if isinstance(before, discord.Message) and before.channel.id != channel.id:
+            raise BotException(
+                "Invalid `before` argument",
+                "`before` has to be an ID to a message from the origin channel",
+            )
+
+        if isinstance(after, discord.Message) and after.channel.id != channel.id:
+            raise BotException(
+                "Invalid `after` argument",
+                "`after` has to be an ID to a message from the origin channel",
+            )
+
+        if isinstance(around, discord.Message) and around.channel.id != channel.id:
+            raise BotException(
+                "Invalid `around` argument",
+                "`around` has to be an ID to a message from the origin channel",
+            )
+
+        if quantity <= 0:
+            if quantity == 0 and not after:
+                raise BotException(
+                    "Invalid `quantity` argument",
+                    "`quantity` must be above 0 when `after=` is not specified.",
+                )
+            elif quantity != 0:
+                raise BotException(
+                    "Invalid `quantity` argument",
+                    "Quantity has to be a positive integer (or `0` when `after=` is specified).",
+                )
+
+        messages = await channel.history(
+            limit=quantity if quantity != 0 else None,
+            before=before,
+            after=after,
+            around=around,
+        ).flatten()
+
+        if not messages:
+            raise BotException(
+                "Invalid time range",
+                "No messages were found for the specified timestamps.",
+            )
+
+        if not after:
+            messages.reverse()
+        self.pages = messages
```

Comment:
In the superclass, self.pages is a list of Embeds. By making it a list of message objects here, it is not super elegant IMHO, rather we could just create all the embeds in this step.

That way, all the code below this line will not be needed as the superclass already handles that for us :wink: